### ISSUE TITLE
object-curly-spacing Does not play nicely with Flow types for function parameter object destructuring

### ISSUE THREAD

**Issue by lemonmade** [babel/babel-eslint#185]
When destructuring an object parameter to a function, it's handy to provide a Flow type to indicate what keys/ types are expected for those options. Here's an example:

```js
type ActiveStyleDetailType = {
  component: string,
  context: ElementType
};

activeStyles(
  styles: Object,
  {component, context}: ActiveStyleDetailType
): Array<CompiledRuleType> { /* ... */ }
```

When the `object-curly-spacing` rule is set to force no spaces inside object literals, the above incorrectly warns that `There should be no space before 'ActiveStyleDetailType'`. Removing the space does resolve this (as does disabling the rule), but both of those prevent us from enforcing our preferred style (no object spacing, <nospace>:<space><FlowType> for Flow annotations).

**Comment by hzoo** [comment]
Will probably need to rewrite the rule in eslint-plugin-babel

**Comment by lemonmade** [comment]
Should I move the issue to that repo?

**Comment by hzoo** [comment]
Can look into it more before doing so. I'm getting ` Unexpected type cast (7:2)` error in from babel using the snippet you provided.

**Comment by tsing** [comment]
I have a snippet which will cause the same problem:

```
type Data = {
  type: string;
  payload: Object
};

export function test({type, payload}: Data) {
  console.log(type, payload);
}
```
eslint warns `There should be no space before 'Data'`

**Comment by hzoo** [comment]
Thanks for the report - we need to move the issue to the other repo since I don't think we can do anything here.
